packages/salesforcedx-vscode-core/src/deploys/toolingDeploy.ts

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { AuthInfo, Connection } from '@salesforce/core';
+import * as fs from 'fs';
+import * as path from 'path';
+import {
+  supportedToolingTypes,
+  ToolingCreateResult,
+  ToolingRetrieveResult
+} from './index';
+// tslint:disable-next-line:no-var-requires
+const DOMParser = require('xmldom-sfdx-encoding').DOMParser;
+
+export class ToolingDeploy {
+  private static instance: ToolingDeploy;
+  private metadataType?: string;
+  private connection?: Connection;
+  private apiVersion?: string;
+
+  private container?: ToolingCreateResult;
+  private containerMember?: ToolingCreateResult;
+  private asyncRequest?: ToolingCreateResult;
```

Comment:
Is there a reason we can't pass these things (along with `metadataType`) as arguments to the methods that need them, instead of having them as properties? Keeping them as optional properties is forcing the use of the non-null assertion (`!` operator) quite a bit which is mitigating the usefulness of typescript keeping us safe.